pgjdbc/src/main/java/org/postgresql/core/AuthenticationPluginManager.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2021, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.core;
+
+import org.postgresql.PGProperty;
+import org.postgresql.plugin.AuthenticationPlugin;
+import org.postgresql.plugin.AuthenticationRequestType;
+import org.postgresql.util.GT;
+import org.postgresql.util.ObjectFactory;
+import org.postgresql.util.PSQLException;
+import org.postgresql.util.PSQLState;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class AuthenticationPluginManager {
+  private static final Logger LOGGER = Logger.getLogger(AuthenticationPluginManager.class.getName());
+
+  private AuthenticationPluginManager() {
+  }
+
+  /**
+   * If a password is requested by the server during connection initiation, this
+   * method will be invoked to supply the password. This method will only be
+   * invoked if the server actually requests a password, e.g. trust authentication
+   * will skip it entirely.
+   *
+   * @param type The authentication type that is being requested
+   * @param info The connection properties for the connection
+   * @return The password to use for authentication or null if none is available
+   * @throws PSQLException Throws a PSQLException if the plugin class cannot be instantiated
+   */
+  public static @Nullable String getPassword(AuthenticationRequestType type, Properties info) throws PSQLException {
+    String authPluginClassName = PGProperty.AUTHENTICATION_PLUGIN_CLASS_NAME.get(info);
+
+    if (authPluginClassName == null || authPluginClassName.equals("")) {
+      // Default auth plugin simply pulls password directly from connection properties
+      return PGProperty.PASSWORD.get(info);
+    }
+
+    AuthenticationPlugin authPlugin;
+    try {
+      authPlugin = (AuthenticationPlugin) ObjectFactory.instantiate(authPluginClassName, info,
```

Comment:
Does it make sense to cache the instance ? Is it possible that different connections would use different authentication plugins ?

Reply:
No it should not be cached as connections should be isolated from each other. If a user needs to maintain state across multiple runtime instantiations then it would need to be done in the user's implementation through either static fields or some kind of proxying.

Reply:
Thinking about that a bit more, perhaps we need a new opaque "arg" property that users can use to customize their implementation. The plugin constructor takes the connection properties but there's no specific property that user's could use for their own purposes. Something like `authPluginArg` (e.g. the plugin equivalent of `socketFactoryArg`).

Reply:
Fairly good chance that the user will only use one auth plugin. I wonder what the cost of instantiating a class really is (in other words is it worth worrying about)

Reply:
Yes I doubt it's an issue in practice and connections tend to be long lived anyway. The time spent opening the underlying socket and TLS handshake is going to be significantly more than any object creation.

Reply:
Fair enough.